openreview/matching.py

```diff
@@ -80,3 +82,294 @@ def get_domains(entity, subdomains = False):
+            user_index = index_by_user.get(user, None)
+            hard_constraint_value = get_hard_constraint_value(user_scores.values())
+
+            if user_index:
+                coordinates = (user_index, paper_index)
+                if hard_constraint_value == -1:
+                    mean = np.mean(user_scores.values()) if user_scores.values() else 0.0
+                    if np.isnan(mean).any():
+                        print forum, user, user_scores
+
+                    score_matrix[coordinates] = mean
+                else:
+                    hard_constraint_dict[coordinates] = hard_constraint_value
+
+    return score_matrix, hard_constraint_dict, user_by_index, forum_by_index
+
+def decode_score_matrix(solution, user_by_index, forum_by_index):
+    '''
+    Decodes the 2D score matrix into a returned dict of user IDs keyed by forum ID.
+
+    e.g. {
+        'abcXYZ': '~Melisa_Bok1',
+        '123-AZ': '~Michael_Spector1'
+    }
+    '''
+
+    assignments_by_forum = defaultdict(list)
+    for var_name in solution:
+        var_val = var_name.split('x_')[1].split(',')
+
+        user_index, paper_index = (int(var_val[0]), int(var_val[1]))
+        user_id = user_by_index[user_index]
+        forum = forum_by_index[paper_index]
+        match = solution[var_name]
+
+        if match == 1:
+            assignments_by_forum[forum].append(user_id)
+
+    return assignments_by_forum
+
+def create_or_update_assignments(assignments, existing_assignments, assignment_invitation, configuration_note, scores_by_user_by_forum):
+    '''
+    Creates or updates (as applicable) the assignment notes with new assignments.
+
+    Returns a list of openreview.Note objects.
+    '''
+
+    alternates = configuration_note.content['configuration']['alternates']
+    label = configuration_note.content['label']
+    new_assignment_notes = []
+    for forum, userids in assignments.iteritems():
+        scores_by_user = scores_by_user_by_forum[forum]
+        assignment = existing_assignments.get(forum, create_assignment(forum, label, assignment_invitation))
+
+        new_content = {
+            'assignedGroups': get_assigned_groups(userids, scores_by_user),
+            'alternateGroups': get_alternate_groups(userids, scores_by_user, alternates)
+        }
+
+        assignment['content'].update(new_content)
+
+        new_assignment_notes.append(openreview.Note(**assignment))
+
+    return new_assignment_notes
+
+def get_assigned_groups(userids, scores_by_user):
+    '''
+    Returns a list of assignment group entries.
+
+    Entries are dicts with the following fields:
+        'finalScore'
+        'scores'
+        'userId'
+    '''
+
+    assignment_entries = []
+    for user, scores in scores_by_user.iteritems():
+        if user in userids:
+            assignment_entries.append({
+                'finalScore': np.mean([val for feature, val in scores.iteritems() if val != '+inf' and val != '-inf']),
```

Comment:
Also here, you are calculating the score again. We should read the value from line L258